### DISCUSSION TITLE
Lazy JWT validation

### DISCUSSION THREAD

**Discussion by rbaerisch** [google-pay/save-to-google-pay-button#12] [Q&A]
Hey there :)

I just started playing around with the button and was trying (but failed) to figure out how to make the button behave "lazy".
Currently the button seems to validate the provided JWT immediately.
What I mean with "lazy" is to have some ability to make the button working only if some users shows the intent to actually use it.

like:
1. immediately render the button but don't provide any JWT yet
2. user shows intent by clicking on it
3. call some async function which will provide the JWT (and eventually will just create the requested Object on the fly)
4. validate & proceed with the provided JWT

Currently (if I got it right) you always need to have some valid JWT when rendering the button - which means we always need to insert the Objects into GPay even if not knowing if it will be requested or not.

Is such a behaviour expected to work somehow?

Thanks Rico

**Comment by JlUgia** [comment]
Hi @rbaerisch!
As you correctly point out the button expects a JWT at construction time. There are underlying items to unfold:
1. Making the button behave asynchronously does not imply much technical wiring. `ButtonOptions` need to be updated to allow an optional `jwt`, and update/extend the component for the framework/library you are using to capture the click event and run an async operation that fetches the JWT from a remote location, to then trigger the save flow.
2. Whether it's valuable to support this flow in the library depends on whether this behavior is not disencouraged or not recommended in the brand and button guidelines (which I'm double-checking as we speak), and the number of users of the library that support a similar asynchronous (if that's not the case, a little example/gist may be more useful).

@socsieng feel free to add to this.
I'll update the thread with the feedback on the flow.

**Reply by socsieng** [reply]
I think it would be a good idea to support it asynchronously, the [underlying API](https://developers.google.com/pay/passes/reference/s2w-reference#gapi.savetoandroidpay.render-function) appears to support some ability to manipulate the `jwt` after the button is clicked: `onprovidejwt`:

> String name of the provide JWT handler function. The purpose of this function is to intercept and potentially manipulate the JWT data before the object is saved into Google Pay. This function takes in no parameters and **must return the JWT as String**. When implementing the event handler, the original JWT data can be retrieved under field `this.getOpenParams().renderData.userParams.jwt`.

However, it's not clear to me if it supports returning `Promise<string>` which I would expect to be required in order to fetch the `jwt` on click, or what kind of parameter needs to be applied to `jwt` initially. Can it be undefined when `onprovidejwt` is used? @JlUgia, is this something you can look into?

Another consideration with creating the `jwt` on click is the added latency introduced on user interaction and whether or not this risks users abandoning the save flow.

**Reply by socsieng** [reply]
From my testing so far:

> it's not clear to me if it supports returning `Promise<string>` which I would expect to be required in order to fetch the `jwt` on click

I does not support returning `Promise<string>` so making an asynchronous call from `onprovidejwt` will be difficult.

> what kind of parameter needs to be applied to jwt initially. Can it be undefined when `onprovidejwt` is used?

Even when `onprovidejwt` is used, `jwt` is still required to be provided by initially. You _may_ be able to provide a valid `jwt` while excluding an object identifier (needs to be confirmed, I haven't tested)
